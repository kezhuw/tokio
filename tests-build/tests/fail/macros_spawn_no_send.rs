use std::rc::Rc;

use tests_build::tokio;

async fn send_f() {}

#[tokio::test(spawn = true)]
async fn no_send_test() {
    let _value = Rc::new(0);
    send_f().await;
}

#[tokio::main(spawn = true)]
async fn no_send_main() {
    let _value = Rc::new(0);
    send_f().await;
}

fn main() {}
