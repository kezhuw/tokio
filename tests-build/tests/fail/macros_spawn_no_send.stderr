error: future cannot be sent between threads safely
  --> tests/fail/macros_spawn_no_send.rs:13:1
   |
13 | #[tokio::main(spawn = true)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ future created by async block is not `Send`
   |
   = help: within `[async block@$DIR/tests/fail/macros_spawn_no_send.rs:13:1: 13:29]`, the trait `Send` is not implemented for `Rc<i32>`
note: future is not `Send` as this value is used across an await
  --> tests/fail/macros_spawn_no_send.rs:16:14
   |
15 |     let _value = Rc::new(0);
   |         ------ has type `Rc<i32>` which is not `Send`
16 |     send_f().await;
   |              ^^^^^ await occurs here, with `_value` maybe used later
17 | }
   | - `_value` is later dropped here
note: required by a bound in `Runtime::spawn`
  --> $WORKSPACE/tokio/src/runtime/runtime.rs
   |
   |     pub fn spawn<F>(&self, future: F) -> JoinHandle<F::Output>
   |            ----- required by a bound in this associated function
   |     where
   |         F: Future + Send + 'static,
   |                     ^^^^ required by this bound in `Runtime::spawn`
